src/containerbuddy/ips.go

```diff
@@ -0,0 +1,282 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"net"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+// GetIP determines the IP address of the container
+func GetIP(specList []string) (string, error) {
+
+	if specList == nil || len(specList) == 0 {
+		// Use a sane default
+		specList = []string{"eth0:inet"}
+	}
+
+	specs, err := parseInterfaceSpecs(specList)
+	if err != nil {
+		return "", err
+	}
+
+	interfaces, interfacesErr := net.Interfaces()
+
+	if interfacesErr != nil {
+		return "", interfacesErr
+	}
+
+	interfaceIPs, interfaceIPsErr := getinterfaceIPs(interfaces)
+
+	/* We had an error and there were no interfaces returned, this is clearly
+	 * an error state. */
+	if interfaceIPsErr != nil && len(interfaceIPs) < 1 {
+		return "", interfaceIPsErr
+	}
+	/* We had error(s) and there were interfaces returned, this is potentially
+	 * recoverable. Let's pass on the parsed interfaces and log the error
+	 * state. */
+	if interfaceIPsErr != nil && len(interfaceIPs) > 0 {
+		log.Printf("We had a problem reading information about some network "+
+			"interfaces. If everything works, it is safe to ignore this "+
+			"message. Details:\n%s\n", interfaceIPsErr)
+	}
+
+	return findIPWithSpecs(specs, interfaceIPs)
+}
+
+// findIPWithSpecs will use the given interface specification list and will
+// find the first IP in the interfaceIPs that matches a spec
+func findIPWithSpecs(specs []interfaceSpec, interfaceIPs []interfaceIP) (string, error) {
+	// Find the interface matching the name given
+	for _, spec := range specs {
+		index := 0
+		iface := ""
+		for _, iip := range interfaceIPs {
+			// Since the interfaces are ordered by name
+			// a change in interface name can safely reset the index
+			if iface != iip.Name {
+				index = 0
+				iface = iip.Name
+			} else {
+				index++
+			}
+			if spec.Match(index, iip) {
+				return iip.IPString(), nil
+			}
+		}
+	}
+
+	// Interface not found, return error
+	return "", fmt.Errorf("None of the interface specifications were able to match\nSpecifications: %s\nInterfaces IPs: %s",
+		specs, interfaceIPs)
+}
+
+// Interface Spec
+type interfaceSpec interface {
+	Match(index int, iip interfaceIP) bool
+	String() string
```

Comment:
Yeah, but we only care about that if the `fmt` is user-visible, in which case we could just pass `someSpecInstance.Spec` to the formatter.